// @rcade/plugin-input-spinners/dist/index.js
// ../../../../node_modules/.pnpm/@rcade+sdk@0.2.2_typescript@5.9.3/node_modules/@rcade/sdk/dist/index.js
var PluginChannel = class _PluginChannel {
  port;
  channel;
  pendingRequests = /* @__PURE__ */ new Map();
  static async acquire(name, version) {
    const nonce = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
    return new Promise((resolve, reject) => {
      const listener = (event) => {
        if (event.data?.type === "plugin_channel" && event.data?.nonce === nonce) {
          const port = event.ports[0];
          if ("error" in event.data) {
            window.removeEventListener("message", listener);
            reject(new Error(event.data.error));
            return;
          }
          const { channel } = event.data;
          if (!channel?.name || !channel?.version || !port) {
            console.warn("Invalid plugin_channel event", event.data);
            return;
          }
          window.removeEventListener("message", listener);
          const pluginChannel = new _PluginChannel(port, channel);
          resolve(pluginChannel);
        }
      };
      window.addEventListener("message", listener);
      window.parent.postMessage({
        type: "acquire_plugin_channel",
        nonce,
        channel: { name, version }
      }, "*");
    });
  }
  constructor(port, channel) {
    this.port = port;
    this.channel = channel;
    this.port.addEventListener("message", (event) => {
      const { _nonce } = event.data ?? {};
      if (_nonce && this.pendingRequests.has(_nonce)) {
        const pending = this.pendingRequests.get(_nonce);
        this.pendingRequests.delete(_nonce);
        pending.resolve(event.data);
      }
    });
    this.port.start();
  }
  getPort() {
    return this.port;
  }
  getVersion() {
    return String(this.channel.version);
  }
  request(message, timeoutMs = 5e3) {
    const _nonce = Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(_nonce);
        reject(new Error("Request timed out"));
      }, timeoutMs);
      this.pendingRequests.set(_nonce, {
        resolve: (data) => {
          clearTimeout(timeout);
          resolve(data);
        },
        reject: (error) => {
          clearTimeout(timeout);
          reject(error);
        }
      });
      this.port.postMessage({ ...message, _nonce });
    });
  }
};

// index.ts
var MAX_DELTA = 1e3;
var stepResolution = 64;
var Spinner = class {
  _stepDelta = 0;
  _angle = 0;
  /** Returns accumulated step delta since last call, then resets to 0. */
  consume_step_delta() {
    const d = this._stepDelta;
    this._stepDelta = 0;
    return d;
  }
  get step_resolution() {
    return stepResolution;
  }
  get angle() {
    return this._angle;
  }
  reset() {
    this._angle = 0;
  }
  /** @internal */
  _update(delta) {
    this._stepDelta = Math.max(-MAX_DELTA, Math.min(MAX_DELTA, this._stepDelta + delta));
    this._angle = Math.atan2(
      Math.sin(this._angle + delta / stepResolution * 2 * Math.PI),
      Math.cos(this._angle + delta / stepResolution * 2 * Math.PI)
    );
  }
};
var spinner1 = new Spinner();
var spinner2 = new Spinner();
var PLAYER_1 = { SPINNER: spinner1 };
var PLAYER_2 = { SPINNER: spinner2 };
var STATUS = { connected: false };
var currentPort = null;
var currentMessageHandler = null;
function cleanup() {
  if (currentPort && currentMessageHandler) {
    currentPort.removeEventListener("message", currentMessageHandler);
    currentPort = null;
    currentMessageHandler = null;
  }
}
(async () => {
  cleanup();
  const channel = await PluginChannel.acquire("@rcade/input-spinners", "^1.0.0");
  currentPort = channel.getPort();
  currentMessageHandler = (event) => {
    const { type, spinner1_step_delta, spinner2_step_delta } = event.data;
    if (type === "spinners") {
      if (spinner1_step_delta !== 0) spinner1._update(spinner1_step_delta);
      if (spinner2_step_delta !== 0) spinner2._update(spinner2_step_delta);
    }
  };
  currentPort.addEventListener("message", currentMessageHandler);
  const config = await channel.request({ type: "get_config" });
  stepResolution = config.step_resolution;
  STATUS.connected = true;
})();
if (import.meta.hot) {
  import.meta.hot.dispose(() => {
    cleanup();
  });
}
export {
  PLAYER_1,
  PLAYER_2,
  STATUS
};
